### PR TITLE
Improve fmtstr_payload()

### PR DESCRIPTION
**Pull request by io12**
This PR improves upon the limitations of `fmtstr_payload()`. The old code doesn't work if any of the wheres has a null byte, because `printf()` stops printing before it reaches the format specifiers. This new code checks if any of the wheres has `\x00`, and places them *after* the format specifiers if they do.

Issue mentioned here: https://github.com/Gallopsled/pwntools/issues/1139

### CODE REVIEW

File: pwnlib/fmtstr.py [hunk 1]
```diff
@@ -151,33 +174,36 @@ def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte'):
...
-    return payload
+            packed_addrs += pack(where+i)
+
+    # decide whether the packed addrs should be placed at the beginning or end
+    # of the payload
+    if '\x00' in packed_addrs:
+        # place at end of payload
+        pad_chr = ' '
+        pad_len = 0 # length to pad the format section to, in processor words
+        while True:
```
Comment (zachriggle) [inline pwnlib/fmtstr.py:196]: Can we make this something like `while (pad_len * context.bytes) < len(fmts)`?

File: pwnlib/fmtstr.py [hunk 2]
```diff
@@ -151,33 +174,36 @@ def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte'):
...
+            # try increasing amounts of padding
+            pad_len += 1
+            fmts = fmts_with_pad(numbwritten, pad_len)
+            if pad_len * context.bytes >= len(fmts):
+                break
+        return fmts.ljust(pad_len * context.bytes, pad_chr) + packed_addrs
```
Comment (zachriggle) [inline pwnlib/fmtstr.py:202]: Tiny nit: I would prefer that we just update `packed_addrs` and have a single return statement at the end.

File: pwnlib/fmtstr.py [hunk 3]
```diff
@@ -151,33 +174,36 @@ def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte'):
     if write_size not in ['byte', 'short', 'int']:
         log.error("write_size must be 'byte', 'short' or 'int'")
 
+    # number: number of write_size in a word
+    # step: amount of bytes in write_size
+    # mask: mask for written values
+    # formatz: format size modifier for %n
+    # declage: amount of bits in write_size
     number, step, mask, formatz, decalage = config[context.bits][write_size]
```
Comment (zachriggle) [inline pwnlib/fmtstr.py:182]: Thanks for adding comments here! The existing code was a bit hard to follow without these

**Review by zachriggle — CHANGES_REQUESTED** [review]

**Comment by Zoruk** [comment]
I was thinking of doing the same patch, this is almost indispensable for 64 bits elf since all addresses starts with a `0x00`. Good job thanks.

File: pwnlib/fmtstr.py (hunk 2 above)
Comment (io12) [reply pwnlib/fmtstr.py:202]: Can we create a new variable `payload` for this? It might be confusing to have `packed_addrs` store the entire payload and not just the packed addresses.

File: pwnlib/fmtstr.py (hunk 1 above)
Comment (io12) [reply pwnlib/fmtstr.py:196]: The issue is `fmts` doesn't exist at that point.

File: pwnlib/fmtstr.py (hunk 2 above)
Comment (io12) [reply pwnlib/fmtstr.py:202]: Either way, I mistakenly deleted the fork for this PR so I don't think I can push to it anymore.

**Comment by Arinerron** [comment]
Is this going to be merged soon? This is a really useful feature.

**Comment by Arinerron** [comment]
@Arusekk Was this feature added in #1322?

**Comment by Arusekk** [comment]
This was resolved in #1216
